2nd alternative fix: allow to change or delete a relyingPartySecret on IdP

strehle opened the pull request:
Alternative to PR: https://github.com/cloudfoundry/uaa/pull/2882, https://github.com/cloudfoundry/uaa/pull/2887
and PR: https://github.com/cloudfoundry/uaa/pull/2885

Only one call in backend, but more input on client side.

authMethod is used to make this change non-breakable.
Because before the update prevented the removal of the relyingPartySecret.
No if authMethod is not client_secret_basic or client_secret_post, then the relyingPartySecret can be overwritten with null.

strehle committed:
WIP: idp secret

strehle committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/idp-secret

strehle committed:
feature: delete secret on existing IdP

- allow to delete a relyingPartySecret on IdP
- Filter IdP list by origin
- Return the auth_method to show current configured client authentication method

strehle committed:
Documentation

strehle committed:
fix names

strehle committed:
sonar

strehle committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/idp-secret

strehle committed:
sonar

strehle committed:
Add patch call to change a secret from an external IdP

strehle committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/idp-secret

strehle committed:
Alias handling

strehle committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/auth_method

strehle committed:
2nd alternative fix: allow to change or delete a relyingPartySecret on IdP

Alternative to PR: #2882, #2887
and PR: #2885

Only one call in backend, but more input on client side.

authMethod is used to make this change non-breakable.
Because before the update prevented the removal of the relyingPartySecret.
No if authMethod is not client_secret_basic or client_secret_post, then the relyingPartySecret can be overwritten with null.

strehle committed:
sonar

strehle committed:
Tests added

strehle committed:
Sonar smell

strehle commented:
sonar fixes

https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2896

strehle committed:
Review

* more checks for edge cases
* more tests to cover edge cases

strehle committed:
Review

Again fixed an edge case
CLIENT_SECRET_BASIC, CLIENT_SECRET_POST both same method, therefore treat them equal

strehle committed:
Review

Again fixed an edge case
Found during tests

strehle committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/auth_method

strehle committed:
Test refactored

peterhaochen47 commented on uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java:
```diff
@@ -591,6 +590,7 @@ void createOAuthIdentityProvider() throws Exception {
                 fieldWithPath("config.userInfoUrl").optional(null).type(STRING).description("A URL for fetching user info attributes when queried with the obtained token authorization."),
                 fieldWithPath("config.showLinkText").optional(true).type(BOOLEAN).description("A flag controlling whether a link to this provider's login will be shown on the UAA login page"),
                 fieldWithPath("config.linkText").optional(null).type(STRING).description("Text to use for the login link to the provider"),
+                fieldWithPath("config.auth_method").optional(null).type(STRING).description("<small><mark>UAA 77.10.0</mark></small> Client authentication method. Possible strings are: client_secret_basic, client_secret_post, private_key_jwt, none."),
```
Why does this field exist in addition to `config.authMethod` (line 607, currently).

strehle replied:
After diff of the whole code I noticed yes, this is a duplicate. We use authMethod not auth_method.

Removed auth_method

peterhaochen47 committed:
small doc fixes

- some clarification & formatting
- no need to call out what the default is in the description
because the `.optional("client_secret_basic")` syntax would
automatically add that language.

peterhaochen47 committed:
doc: clarify when external OIDC client auth requirements

- clarify when config.jwtClientAuthentication and config.relyingPartySecret
are required in relation to the new field config.authMethod

strehle committed:
Review, removed auth_method which is not used, but we ue authMethod field only

strehle committed:
revert this

strehle committed:
remove deprecated
